### DISCUSSION TITLE
[RFC] HTTP Middleware

### DISCUSSION THREAD

**Discussion by flybayer** [blitz-js/blitz#326] [RFCs] [answered]
We already proposed the middleware API in the [architecture RFC](https://github.com/blitz-js/blitz/blob/rfc-architecture/rfc-docs/01-architecture.md#f-middleware), but we need to flesh out a few more details before building.

Middleware is very high priority since it's a blocker for auth, so let's hash this out quickly.

### `preMiddleware`

- Runs before a query/mutation
- Takes the raw Node req & res objects as well as current context
- Can modify the context and return it (just like the webpack override works in blitz.config.js)
- Context is empty object by default unless a middleware adds to it

### `postMiddleware`

- Takes same args as `preMiddleware` with addition of query/mutation result
- Can modify the raw request if needed
- Return type is void

```tsx
// /app/queries/getReferrer.ts
import {PreMiddlewareInput, PostMiddlewareInput} from 'blitz/types'

type ReferrerContext = {referrer: string}
type SpecialContext = ReferrerContext /* | OtherCustomContext */


export const preMiddleware = [
  ({req, ctx}: PreMiddlewareInput): ReferrerContext => {
    ctx.referrer = req.headers.referer
    return {
      ctx
    }
  },
]
export const postMiddleware = [
  ({res}: PostMiddlewareInput<typeof getReferrer>) => {
    res.statusCode(418)
  },
]

export default async function getReferrer(data: any, ctx?: SpecialContext) {
  return ctx?.referrer
}
```

### Global Query/Mutation Middleware

Can add middleware to `blitz.config.js` for it to run for all queries/mutations.

```js
// blitz.config.js
const authMiddleware = require('./somewhere')

module.exports = {
  preMiddleware: [authMiddleware],
  webpack: (config) => {
    return config
  },
}

```

### Middleware Priority

Do we need a way to specify middleware priority? I.e. which runs before another?

By default global middleware would run first in array order, then local query/mutation middleware would run next, also in array order.

---

cc @ryardley @merelinguist @rishabhpoddar

**Comment by dajinchu** [comment]
Curious if it would make sense to have a way to specify middleware for a whole folder/page? I'm imagining with an authentication app there's commonly unauthenticated and authenticated pages, so you wouldn't want to add auth middleware globally, but it might be a pain to add it to each individual endpoint.

**Comment by flybayer** [comment]
Good point @dajinchu. Pages will be authenticated at the component level, so middleware won't directly affect that.

I think authentication is fine as a global middleware because authorization will be handled inside queries/mutations. The so if a user isn't logged in but needs to be, the query will throw the Authorization error, not the middleware.

Folder level middleware could easily be added later. So it's not critical for now, but definitely good to consider this.

**Comment by ryardley** [comment]
You might be thinking this just not sure if this is clear here but we may want to short circuit the response in the middleware so we should ensure we have access to both req, res and next just like connect middleware.

**Comment by ryardley** [comment]
Maybe this is weird but I think it would be good to be able to put middleware on a query by query basis too like express. Not sure how to do this syntactically ( maybe decorators?)

**Comment by ryardley** [comment]
Last thought from me: is there any way we can remain compatible with connect middleware for this or is that crazy talk/a dumb idea? `req.locals ->  resolverCtx` might make it easy to implement using `next-connect`?

**Comment by flybayer** [comment]
> You might be thinking this just not sure if this is clear here but we may want to short circuit the response in the middleware so we should ensure we have access to both req, res and next just like connect middleware.

When would it truly be better for middleware to short circuit the request vs in the actual query/mutation?



> Maybe this is weird but I think it would be good to be able to put middleware on a query by query basis too like express. Not sure how to do this syntactically ( maybe decorators?)

If I understand what you are saying, my proposal includes this by exporting middleware from the query/mutation file

> Last thought from me: is there any way we can remain compatible with connect middleware for this or is that crazy talk/a dumb idea? `req.locals -> resolverCtx` might make it easy to implement using `next-connect`?

🤷 I don't know — what does everyone thing about this?

**Comment by ryardley** [comment]
> When would it truly be better for middleware to short circuit the request vs in the actual query/mutation?

Access control / special redirect rules for api routes / cache middleware

> If I understand what you are saying, my proposal includes this by exporting middleware from the query/mutation file

👍

**Comment by flybayer** [comment]
Alright, in absence of any good reason to not be Connect middleware compatible, I think that's probably the correct approach.

Initially I didn't really consider this because of how we are abstracting away the http layer. But I think we'd probably regret not being connect compatible.

I think that would also mean we don't need both `preMiddleware` and `postMiddleware`, right?

[Looks like we can run the middleware ourselves](https://nextjs.org/docs/api-routes/api-middlewares#connectexpress-middleware-support) without `next-connect`.

**Comment by MrLeebo** [comment]
> If I understand what you are saying, my proposal includes this by exporting middleware from the query/mutation file

Doesn't seem composable if you have multiple middlewares that are shared and you want to apply to the query. Perhaps allow the query export to be an array of [middlewareA, middlewareB, queryFn]?

```js
import {Context} from 'blitz/types'
import db, {FindOneProductArgs} from 'db'
import acl from 'blitz-acl-middleware'
import cache from 'blitz-cache-middleware'

async function getProduct(args: FindOneProductArgs, ctx?: Context) {
  // Can do any pre-processing here or trigger events

  const product = await db.setUser(ctx.session.user).product.findOne(args)

  // Can do any post-processing here or trigger events

  return product
}

// export an array of middlewares followed by the query or mutation
export default [acl({ resource: 'products' }), cache({ maxAge: 10 }), getProduct]
```

**Comment by flybayer** [comment]
So remember what we are doing here. We aren't building a traditional REST API that's tightly coupled to HTTP specific stuff. We're building an RPC system where extremely little logic lives in the HTTP layer. This is quite similar to GraphQL. GraphQL and RPC both abstract away the HTTP layer. You never think about an HTTP status code with GraphQL, for example. Both GraphQL and our RPC layer always take a POST request and return the same status code for most responses.

I think HTTP middleware is mostly anti-composable. It's more like mixins where where you don't know exactly what, where, or why something is happening. Placing business logic, like access control, in the HTTP layer makes your queries and mutations much harder to test. They are no longer plain JS functions, but full-on HTTP request handlers.

Maybe I'm totally crazy and missing something obvious (in which case please enlighten me! 😅), but I think 100% of business logic in Blitz apps should live inside query and mutation functions. Not in HTTP middleware.

I view Blitz HTTP middleware as a rare escape hatch for things that are impossible to do inside queries/mutations. The perfect example is session management which needs to read and set cookies. But that's all session management middleware will do. The session object will be passed into queries/mutations in the context object where access control will take place.

That makes queries/mutations very nice to test because it's a plain JS function that accepts a plain session object. Nothing HTTP specific at all is needed. These queries/mutations are also very composable simply by passing along the context object. All without any reliance on HTTP specifics.

Decorators could be a good way to wrap queries/mutations with access control.

If someone needs a more traditional REST approach with lots of HTTP specific stuff, then they can do that with custom API routes.

What do you think?

**Comment by ryardley** [comment]
> Placing business logic, like access control, in the HTTP layer makes your queries and mutations much harder to test. 

Great point

**Comment by dwightwatson** [comment]
I like the idea of using something like `next-connect` where you don't need specific pre-/post- middleware - instead you can adjust how your middleware interacts with the `next()` callback. Makes it easy to configure as a pipeline instead of having to queue up middleware to run before and after a request is handled.

**Comment by flybayer** [comment]
## Update!

I'm working on implementing connect compatible middleware. With three additions:

1. The ability to `await next()` like Koa middleware while remaining connect compatible
2. `res.blitzCtx` which will be passed as the second argument to queries/mutations. 
3. `res.blitzResult` which will be the result returned from the query/mutation



Query/mutation execution will itself be a middleware. This will be ran after all custom middleware.

Here's example code I'm working towards:

```tsx
// app/queries/getReferrer.ts
import {Middleware} from 'blitz'

const customMiddleware: Middleware = async (req, res, next) => {
  res.blitzCtx.referrer = req.headers.referer

  await next()

  console.log('Query result': res.blitzResult)  
}

export const middleware = [customMiddleware]

export default async function getReferrer(data, ctx) {
  return ctx?.referrer
}
```

### Global Query/Mutation Middleware

Can add middleware to `blitz.config.js` for it to run for all queries/mutations.

```js
// blitz.config.js
const authMiddleware = require('./somewhere')

module.exports = {
  middleware: [authMiddleware],
  webpack: (config) => {
    return config
  },
}

```

**Comment by flybayer** [comment, marked as answer]
HTTP middleware is added in https://github.com/blitz-js/blitz/pull/652!

**Comment by rockymeza** [comment]
@flybayer I'm a little concerned about the concept of local middleware. I really liked when I saw this comment:

> I think 100% of business logic in Blitz apps should live inside query and mutation functions.

I was wondering how one would go about using queries/mutations from places other than the frontend. For example, if a background worker needs to make a mutation, would it have to call the middleware itself? I suppose it could build out context objects and pass those into the query/mutation function.

Another concern that I have is about the type safety of the query/mutation functions. What is the type of `blitzCtx`? If a background worker is calling a query/mutation, how can we know if the shape of the context that we build is correct?

Thanks!

**Reply by flybayer** [reply]
> if a background worker needs to make a mutation, would it have to call the middleware itself? 

This is exactly why middleware should not contain business logic. If middleware is only for HTTP specific things like cookies, then using a mutation from a background worker will not need to use the middleware. It can call the mutation directly and pass in a custom context object if needed.

The type of `blitzCtx` is `Record<any, unknown>`. It's impossible to type this because middleware can modify this in any way without you knowing (but of course middleware authors shouldn't do anything stupid here :)

> If a background worker is calling a query/mutation, how can we know if the shape of the context that we build is correct?

You can add your own types to the second `ctx` parameter to your query/mutation. This type won't cause a compile error if middleware doesn't conform to it, but it does ensure you pass in the correct ctx object from a background worker for example.
